docker.go

```diff
@@ -899,20 +899,25 @@ func (p *DockerProvider) BuildImage(ctx context.Context, img ImageBuildInfo) (st
 
 	var buildError error
 	var resp types.ImageBuildResponse
-	err = backoff.Retry(func() error {
-		resp, err = p.client.ImageBuild(ctx, buildOptions.Context, buildOptions)
-		if err != nil {
-			buildError = errors.Join(buildError, err)
-			if isPermanentClientError(err) {
-				return backoff.Permanent(err)
+	err = backoff.RetryNotify(
+		func() error {
+			resp, err = p.client.ImageBuild(ctx, buildOptions.Context, buildOptions)
+			if err != nil {
+				buildError = errors.Join(buildError, err)
+				if isPermanentClientError(err) {
+					return backoff.Permanent(err)
+				}
+				return err
 			}
-			Logger.Printf("Failed to build image: %s, will retry", err)
```

Comment:
Note that if the error is logged here, the last error will be logged twice: once here and once by the caller when `backoff.RetryNotify()` returns it.

Reply:
Hence, we use the callback version that only calls the notify callback when the error will be retried.

Reply:
TIL